check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java

```diff
@@ -2647,13 +2648,17 @@ private void scan(Type from, Type to) {
   /** Returns {@code true} if this is a `var` or a lambda parameter that has no explicit type. */
   public static boolean hasNoExplicitType(VariableTree tree, VisitorState state) {
     /*
-     * We detect the absence of an explicit type by looking for an absent start position for the
-     * type tree.
-     *
      * Note that the .isImplicitlyTyped() method on JCVariableDecl returns the wrong answer after
      * type attribution has occurred.
      */
-    return getStartPosition(tree.getType()) == -1;
+    return !hasExplicitSource(tree.getType(), state);
```

Comment:
We used to check `getEndPosition` in `hasNoExplicitType` to work around a javac8 bug, and the workaround was removed in https://github.com/google/error-prone/commit/1b8bf175843e3acaa6f22aa9c3070607af103273. I can't remember if there was any motivation for that beyond removing references to javac8 bugs after we dropped JDK 8 support.

I also remembered that recent versions of `JCVariableDecl` (including in JDK 11u) have a `declaredUsingVar` getter that I think we could be using here. We might also need to check `isImplicitlyTyped`.

Thoughts on using that approach instead of delegating? I think it reads as well and avoids having to check end positions.

Reply:
Though the unconditional downcast casts always make me a little uncomfortable, there's precedent even in this class to unconditionally cast to `JCVariableDecl`. So I tried this:
```java
public static boolean hasImplicitType(VariableTree tree, VisitorState state) {
  JCVariableDecl varTree = (JCVariableDecl) tree;
  return varTree.declaredUsingVar() || varTree.isImplicitlyTyped();
}
```

Unfortunately this causes the existing `negativeLambdaParameterNoType` test to fail:
```
java.lang.AssertionError: Saw unexpected error on line 5. All errors:
/Test.java:5: Note: [VoidMissingNullable] The type Void is not annotated @Nullable
  Test TEST = v -> {};
              ^
    (see https://errorprone.info/bugpattern/VoidMissingNullable)
  Did you mean 'Test TEST = @Nullable v -> {};'?

	at org.junit.Assert.fail(Assert.java:89)
	at com.google.errorprone.DiagnosticTestHelper.assertHasDiagnosticOnAllMatchingLines(DiagnosticTestHelper.java:348)
	at com.google.errorprone.CompilationTestHelper.doTest(CompilationTestHelper.java:342)
	at com.google.errorprone.bugpatterns.nullness.VoidMissingNullableTest.negativeLambdaParameterNoType(VoidMissingNullableTest.java:294)
	...
```

When I attach a debugger then I see that indeed `declaredUsingVar` is false, and `vartype` is a `JCIdent` representing `Void`. I'm guessing (but did not investigate deeper) that this has something to do with the lambda expression being assigned the type of the targeted functional interface.

I'll propose a new comment to document this observation.

(Tested with Temurin-11.0.20.1+1 and Temurin-17.0.8.1+1.)

Reply:
Thanks for trying that out, and for the comment.